app/handlers/base.py

```diff
@@ -86,6 +86,48 @@ def log_exception(self, typ=None, value=None, tb=None):
             )
 
 
+def bulk_verify(mode, collection, accessor):
+    """Vectorized permission check for a heterogeneous set of records. If an
+    access leak is detected, it will be handled according to the `security`
+    section of the application's configuration.
+
+    Parameters
+    ----------
+    mode: str
+        The access mode to check. Can be create, read, update, or delete.
+    collection: collection of `baselayer.app.models.Base`.
+        The records to check. These records will be grouped by type, and
+        a single database query will be issued to check access for each
+        record type.
+    accessor: baselayer.app.models.User or baselayer.app.models.Token
+        The user or token to check.
+    """
+
+    grouped_collection = defaultdict(list)
+    for row in collection:
+        grouped_collection[type(row)].append(row)
+
+    # check all rows of the same type with a single database query
+    for record_cls, collection in grouped_collection.items():
+        collection_ids = set(record.id for record in collection)
+
+        # vectorized query for ids of rows in the session that
+        # are accessible
+        accessible_row_ids = record_cls.query_records_accessible_by(
+            accessor, mode=mode, columns=[record_cls.id]
+        ).filter(record_cls.id.in_(collection_ids))
+
+        # compare the accessible ids with the ids that are in the session
+        accessible_row_ids = set(id for id, in accessible_row_ids)
```

Comment:
Wow, I actually never realized that SQLAlchemy Query instances were iterators themselves - thought you had to call `.all()` every time. Good to know!